Bio/PDB/MMCIFIO.py

```diff
@@ -0,0 +1,301 @@
+                        except Exception:
+                            pass
+                        sb.structure[0]['A'].add(pdb_object)
+                    else:
+                        # Atom
+                        sb.init_residue('DUM', ' ', 1, ' ')
+                        try:
+                            parent_id = pdb_object.parent.parent.id
+                            sb.structure[0]['A'].id = parent_id
+                        except Exception:
+                            pass
+                        sb.structure[0]['A'].child_list[0].add(pdb_object)
+
+            # Return structure
+            structure = sb.structure
+        self.structure = structure
+
+    def set_dict(self, dic):
+        # Set the mmCIF dictionary to be written out
+        self.dic = dic
+
+    def save(self, filepath, select=Select(), preserve_atom_numbering=False):
+        if isinstance(filepath, basestring):
+            fp = open(filepath, "w")
+            close_file = True
+        else:
+            fp = filepath
+            close_file = False
+        # Decide whether to save a Structure object or an mmCIF dictionary
+        if hasattr(self, 'structure'):
+            self._save_structure(fp, select, preserve_atom_numbering)
+        elif hasattr(self, 'dic'):
+            self._save_dict(fp)
+        else:
+            raise ValueError("Use set_structure or set_dict to set a structure or dictionary to write out")
+        if close_file:
+            fp.close()
+
+    def _save_dict(self, out_file):
+        # Form dictionary where key is first part of mmCIF key and value is list
+        # of corresponding second parts
+        key_lists = {}
+        for key in self.dic:
+            if key == "data_":
+                data_val = self.dic[key]
+            else:
+                s = re.split(r"\.", key)
+                if len(s) == 2:
+                    if s[0] in key_lists:
+                        key_lists[s[0]].append(s[1])
+                    else:
+                        key_lists[s[0]] = [s[1]]
+                else:
+                    raise ValueError("Invalid key in mmCIF dictionary: " + key)
+
+        # Re-order lists if an order has been specified
+        # Not all elements from the specified order are necessarily present
+        for key, key_list in key_lists.items():
+            if key in mmcif_order:
+                inds = []
+                for i in key_list:
+                    try:
+                        inds.append(mmcif_order[key].index(i))
+                    # Unrecognised key - add at end
+                    except ValueError:
+                        inds.append(len(mmcif_order[key]))
+                z = list(zip(inds, key_list))
+                z.sort()
+                key_lists[key] = [k for _, k in z]
+
+        # Write out top data_ line
+        if data_val:
+            out_file.write("data_" + data_val + "\n")
+            out_file.write("#\n")
```

Comment:
If disk IO is going to be slow with mmCIF files, you might as well combine these two lines into a single call to write:

```python
out_file.write("data_" + data_val + "\n#\n")
```

If performance is not critical, then your version is much clearer.